storage/innobase/row/row0mysql.cc

```diff
@@ -228,24 +279,345 @@ row_mysql_read_true_varchar(
+		prebuilt->compress_heap =
+			mem_heap_create(max(UNIV_PAGE_SIZE, buf_len));
+	}
+
+	buf = static_cast<byte*>(mem_heap_zalloc(
+			prebuilt->compress_heap,buf_len));
+
+	if (*len < srv_compressed_columns_threshold ||
+		srv_compressed_columns_zip_level == Z_NO_COMPRESSION)
+		goto do_not_compress;
+
+	ptr = buf + COLUMN_COMPRESS_HEADER_LEN + lenlen;
+
+	/*init deflate object*/
+	c_stream.next_in = (Bytef*)data;
+	c_stream.avail_in = *len;
+	c_stream.next_out = ptr;
+	c_stream.avail_out = comp_len;
+
+	column_zip_set_alloc(&c_stream, prebuilt->compress_heap);
+
+	err = deflateInit2(&c_stream, srv_compressed_columns_zip_level,
+			Z_DEFLATED, window_bits, MAX_MEM_LEVEL, srv_compressed_columns_zlib_strategy);
+	ut_a(err == Z_OK);
+
+	if(dict_data != 0 && dict_data_len != 0)
+	{
+		err = deflateSetDictionary(&c_stream, dict_data, dict_data_len);
+		ut_a(err == Z_OK);
+	}
+	
+	err = deflate(&c_stream, Z_FINISH);
+	if (err != Z_STREAM_END) {
+		deflateEnd(&c_stream);
+		if (err == Z_OK)
+			err = Z_BUF_ERROR;
+	} else {
+		comp_len = c_stream.total_out;
+		err = deflateEnd(&c_stream);
+	}
+
+	switch(err) {
+	case Z_OK:
+		break;
+	case Z_BUF_ERROR:
+		/* data after compress is larger than uncompressed data*/
+		break;
+	default:
+		ib_logf(IB_LOG_LEVEL_ERROR, "failed to compress the column, error: %d\n", err);
+	}
+
+	/* make sure the compressed data size is smaller than uncmpressed data*/
+	if (err == Z_OK && *len > (comp_len + COLUMN_COMPRESS_HEADER_LEN + lenlen))
+	{
+		column_set_compress_header(buf, 1, lenlen-1, 0, wrap);
+		ptr = buf+1;
+		/*store the uncompressed data length*/
+		switch(lenlen) {
+		case 1:
+			mach_write_to_1(ptr, *len);
+			break;
+		case 2:
+			mach_write_to_2(ptr, *len);
+			break;
+		case 3:
+			mach_write_to_3(ptr, *len);
+			break;
+		case 4:
+			mach_write_to_4(ptr, *len);
+			break;
+		default:
+			ut_a(0);
+		}
+
+		*len = comp_len+COLUMN_COMPRESS_HEADER_LEN+lenlen;
+		return buf;
+	}
+
+do_not_compress:
+	ptr = buf;
+	column_set_compress_header(ptr, 0, 0, 0, 0);
+	ptr += COLUMN_COMPRESS_HEADER_LEN;
+	memcpy(ptr, data, *len);
+	*len += COLUMN_COMPRESS_HEADER_LEN;
+	return buf;
+}
+
+/*******************************************************************//**
+Uncompress column data using zlib
+@return pointer to the uncompressed data */
+const byte*
+row_decompress_column(
+	const byte*	data,		/*!< in: data in innodb(compressed) format */
+	ulint*		len,		/*!< in: data length; out: length of decompressed data*/
+	const byte*	dict_data,	/*!< in: optional dictionary data used for decompression */
+	ulint		dict_data_len,	/*!< in: optional dictionary data length */
+	row_prebuilt_t*	prebuilt)	/*!< in: use prebuilt->compress_heap only here*/
+{
+	ulint buf_len = 0;
+	byte* buf;
+	int err = 0;
+	int window_bits = 0;
+	z_stream d_stream;
+	bool is_compress = false;
+	bool wrap = false;
+	ulint lenlen = 0;
+	uint alg = 0;
+
+	column_get_compress_header(data, &is_compress, &lenlen, &alg, &wrap);
+
+	ut_a(alg == 0);
```

Comment:
Added `"unsupported 'algorithm' value in the compressed BLOB header` message along with the assertion.